Mathlib/Data/List/Destutter.lean

```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+  cases h2 : l with
+  | nil => simp -- l = []
+  | cons a as =>
+    clear h2
+    induction as generalizing a with
+    | nil => simp -- l = a :: []
+    | cons a2 bs ih => -- l = a :: a2 :: bs
+      repeat rw [map_cons, destutter_cons_cons]
+      simp_rw [← h a a2]
+      by_cases hr : (R a a2) <;>
+        simp [hr, ← destutter_cons', ih]
+
+/-- For a injective function `f`, `destutter' (·≠·)` commutes with `map f`. -/
+theorem map_destutter_ne {f : α → β} (h : Injective f) [DecidableEq α] [DecidableEq β] :
+    (l.destutter (·≠·)).map f = (l.map f).destutter (·≠·) :=
+  destutter_map_iff l _ f (fun _ _ ↦ (Injective.ne_iff h).symm)
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
+    the first elements are not related (¬Rα a b). --/
+theorem length_destutter' [IsEquiv α Rᶜ] (hab : ¬R a b) :
+    (List.destutter' R a l).length = (List.destutter' R b l).length := by
+  induction l with
+  | nil => simp
+  | cons c cs ih =>
+    obtain hac | hac : R a c ∨ Rᶜ a c := em _
+    · have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (not_not.2 hac)
+      simp [destutter', if_pos hac, if_pos (not_not.1 hbc)]
+    · have hbc : ¬R b c := trans (symm hab) hac
+      simpa [destutter', if_neg hac, if_neg hbc]
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem le_length_destutter' [IsEquiv α Rᶜ] :
+    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
+  cases l with
+  | nil => by_cases hab : (R a b) <;> simp_all [Nat.le_succ]
+  | cons c cs =>
+    by_cases hab : (R a b)
+    case pos => simp [destutter', if_pos hab, Nat.le_succ]
+    obtain hac | hac : R a c ∨ Rᶜ a c := em _
+    · have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (not_not.2 hac)
+      simp [destutter', if_pos hac, if_pos (not_not.1 hbc), if_neg hab]
+    · have hbc : ¬R b c := trans (symm hab) hac
+      apply le_of_eq
+      simp only [destutter', if_neg hbc, if_neg hac, if_neg hab]
+      exact (length_destutter' cs R hab).symm
+
+/-- `destutter` on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem length_destutter_cons_ge_length_destutter [IsEquiv α Rᶜ] :
+    (l.destutter R).length ≤ ((a::l).destutter R).length := by
+  cases l
+  · simp [destutter]
+  · exact le_length_destutter' _ R
+
+/-- `destutter ≠` has length monotonic under List.cons --/
+theorem length_destutter_ne_cons_ge_length_destutter [DecidableEq α]:
+    (l.destutter (·≠·)).length ≤ ((a::l).destutter (·≠·)).length :=
+  length_destutter_cons_ge_length_destutter l (·≠·)
+
+/-- `destutter'` on a relation like ≠ or <, whose negation is transitive, has length monotonic
+    under a ¬R changing of the first element. -/
+theorem length_destutter'_cotrans_ge [i : IsTrans α Rᶜ] (hba : ¬R b a) :
+    (List.destutter' R b l).length ≤ (List.destutter' R a l).length := by
```

Comment:
Does this used twice not imply ` length_destutter'` straightforwardly?

Reply:
It is, good point. Changed